src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -703,6 +706,60 @@ private static bool IsThrowExpressionInProperContext(ThrowExpressionSyntax node)
             }
         }
 
+        // Bind a declaration expression where it isn't permitted.
+        private BoundExpression BindDeclarationExpression(DeclarationExpressionSyntax node, DiagnosticBag diagnostics)
+        {
+            // This is an error, as declaration expressions are handled specially in every context in which
+            // they are permitted. So we have a context in which they are *not* permitted. Nevertheless, we
+            // bind it and then give one nice message.
+
+            bool isVar;
+            bool isConst = false;
+            AliasSymbol alias;
+            TypeSymbol declType = BindVariableType(node.Designation, diagnostics, node.Type, ref isConst, out isVar, out alias);
+            Error(diagnostics, ErrorCode.ERR_DeclarationExpressionNotPermitted, node);
+            return BindDeclarationVariables(declType, node.Designation, diagnostics);
+        }
+
+        // Bind a declaration variable where it isn't permitted.
+        private BoundExpression BindDeclarationVariables(TypeSymbol declType, VariableDesignationSyntax node, DiagnosticBag diagnostics)
+        {
+            switch (node.Kind())
+            {
+                case SyntaxKind.SingleVariableDesignation:
+                    {
+                        var single = (SingleVariableDesignationSyntax)node;
+                        var result = BindDeconstructionVariable(declType, single, diagnostics);
+
+                        // for error recovery, we force inference of the type of a misplaced declaration expression
+                        if ((object)declType == null)
+                        {
+                            result = SetInferredType(result, CreateErrorType("var"), diagnostics);
+                        }
+                        return result;
+                    }
+                case SyntaxKind.DiscardedDesignation:
+                    {
+                        var discarded = (DiscardedDesignationSyntax)node;
+                        return BindDiscardedExpression(discarded, declType);
+                    }
+                case SyntaxKind.ParenthesizedVariableDesignation:
+                    {
+                        var tuple = (ParenthesizedVariableDesignationSyntax)node;
+                        var builder = ArrayBuilder<BoundExpression>.GetInstance();
+                        foreach (var n in tuple.Variables)
+                        {
+                            builder.Add(BindDeclarationVariables(declType, n, diagnostics));
+                        }
+                        var subExpressions = builder.ToImmutableAndFree();
+                        var tupleType = (TypeSymbol)this.Compilation.CreateTupleTypeSymbol(subExpressions.Select<BoundExpression, ITypeSymbol>(e => e.Type).ToImmutableArray());
```

Comment:
Consider using `TupleTypeSymbol` directly rather than `ITypeSymbol`, etc.:
```
TupleTypeSymbol.Create(
    null,
    subExpressions.SelectAsArray(e => e.Type),
    default(ImmutableArray<Location>),
    default(ImmutableArray<string>,
    Compilation);
```